src/prototype_rust_sdk/workflow_context.rs

```diff
@@ -0,0 +1,130 @@
+use crate::{
+    protos::coresdk::{
+        activity_result::ActivityResult,
+        common::Payload,
+        workflow_commands::{ScheduleActivity, StartTimer},
+    },
+    prototype_rust_sdk::{CommandCreateRequest, RustWfCmd, UnblockEvent},
+};
+use crossbeam::channel::{Receiver, Sender};
+use futures::{task::Context, FutureExt};
+use std::{future::Future, pin::Pin, task::Poll, time::Duration};
+use tokio::sync::{oneshot, watch};
+
+/// Used within workflows to issue commands, get info, etc.
+pub struct WfContext {
+    chan: Sender<RustWfCmd>,
+    args: Vec<Payload>,
+    am_cancelled: watch::Receiver<bool>,
+}
+
+impl WfContext {
+    /// Create a new wf context, returning the context itself, the shared cache for blocked
+    /// commands, and a receiver which outputs commands sent from the workflow.
+    pub(super) fn new(
+        args: Vec<Payload>,
+        am_cancelled: watch::Receiver<bool>,
+    ) -> (Self, Receiver<RustWfCmd>) {
+        // We need to use a normal std channel since our receiving side is non-async
+        let (chan, rx) = crossbeam::channel::unbounded();
+        (
+            Self {
+                chan,
+                args,
+                am_cancelled,
+            },
+            rx,
+        )
+    }
+
+    /// Get the arguments provided to the workflow upon execution start
+    pub fn get_args(&self) -> &[Payload] {
+        self.args.as_slice()
+    }
+
+    /// A future that resolves if/when the workflow is cancelled
+    pub async fn cancelled(&mut self) {
+        if *self.am_cancelled.borrow() {
+            return;
+        }
+        self.am_cancelled
+            .changed()
+            .await
+            .expect("Cancelled send half not dropped")
+    }
+
+    /// Request to create a timer
+    pub fn timer(&mut self, a: StartTimer) -> impl Future<Output = ()> {
+        let (cmd, unblocker) = WFCommandFut::new();
+        self.send(
+            CommandCreateRequest {
+                cmd: a.into(),
+                unblocker,
+            }
+            .into(),
+        );
+        cmd.map(|ue| {
+            if let UnblockEvent::Timer(_) = ue {
+            } else {
+                panic!("Wrong unblock event")
+            }
+        })
+    }
+
+    /// Request to run an activity
+    pub fn activity(&mut self, a: ScheduleActivity) -> impl Future<Output = ActivityResult> {
+        let (cmd, unblocker) = WFCommandFut::new();
+        self.send(
+            CommandCreateRequest {
+                cmd: a.into(),
+                unblocker,
+            }
+            .into(),
+        );
+        cmd.map(|ue| {
+            if let UnblockEvent::Activity { result, .. } = ue {
+                result
+            } else {
+                panic!("Wrong unblock event")
+            }
+        })
+    }
+
+    /// Cancel a timer
+    pub fn cancel_timer(&self, timer_id: &str) {
+        self.send(RustWfCmd::CancelTimer(timer_id.to_string()))
```

Comment:
Eventually you'll want to use context or cancellation scopes for doing cancellations.

Reply:
Yep